// @angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system_windows", ["require", "exports", "tslib", "path", "@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MockFileSystemWindows = void 0;
    var tslib_1 = require("tslib");
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    /// <reference types="node" />
    var p = require("path");
    var mock_file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system/testing/src/mock_file_system");
    var MockFileSystemWindows = /** @class */ (function (_super) {
        tslib_1.__extends(MockFileSystemWindows, _super);
        function MockFileSystemWindows() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        MockFileSystemWindows.prototype.resolve = function () {
            var _a;
            var paths = [];
            for (var _i = 0; _i < arguments.length; _i++) {
                paths[_i] = arguments[_i];
            }
            var resolved = (_a = p.win32).resolve.apply(_a, tslib_1.__spreadArray([this.pwd()], tslib_1.__read(paths)));
            return this.normalize(resolved);
        };
        MockFileSystemWindows.prototype.dirname = function (path) {
            return this.normalize(p.win32.dirname(path));
        };
        MockFileSystemWindows.prototype.join = function (basePath) {
            var _a;
            var paths = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                paths[_i - 1] = arguments[_i];
            }
            return this.normalize((_a = p.win32).join.apply(_a, tslib_1.__spreadArray([basePath], tslib_1.__read(paths))));
        };
        MockFileSystemWindows.prototype.relative = function (from, to) {
            return this.normalize(p.win32.relative(from, to));
        };
        MockFileSystemWindows.prototype.basename = function (filePath, extension) {
            return p.win32.basename(filePath, extension);
        };
        MockFileSystemWindows.prototype.isRooted = function (path) {
            return /^([A-Z]:)?([\\\/]|$)/i.test(path);
        };
        MockFileSystemWindows.prototype.splitPath = function (path) {
            return path.split(/[\\\/]/);
        };
        MockFileSystemWindows.prototype.normalize = function (path) {
            return path.replace(/^[\/\\]/i, 'C:/').replace(/\\/g, '/');
        };
        return MockFileSystemWindows;
    }(mock_file_system_1.MockFileSystem));
    exports.MockFileSystemWindows = MockFileSystemWindows;
});
