// @angular/compiler-cli/src/ngtsc/incremental/src/incremental.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/incremental/src/incremental", ["require", "exports", "tslib", "@angular/compiler-cli/src/ngtsc/file_system", "@angular/compiler-cli/src/ngtsc/perf", "@angular/compiler-cli/src/ngtsc/program_driver", "@angular/compiler-cli/src/ngtsc/util/src/typescript", "@angular/compiler-cli/src/ngtsc/incremental/semantic_graph", "@angular/compiler-cli/src/ngtsc/incremental/src/dependency_tracking", "@angular/compiler-cli/src/ngtsc/incremental/src/state"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.IncrementalCompilation = void 0;
    var tslib_1 = require("tslib");
    var file_system_1 = require("@angular/compiler-cli/src/ngtsc/file_system");
    var perf_1 = require("@angular/compiler-cli/src/ngtsc/perf");
    var program_driver_1 = require("@angular/compiler-cli/src/ngtsc/program_driver");
    var typescript_1 = require("@angular/compiler-cli/src/ngtsc/util/src/typescript");
    var semantic_graph_1 = require("@angular/compiler-cli/src/ngtsc/incremental/semantic_graph");
    var dependency_tracking_1 = require("@angular/compiler-cli/src/ngtsc/incremental/src/dependency_tracking");
    var state_1 = require("@angular/compiler-cli/src/ngtsc/incremental/src/state");
    /**
     * Discriminant of the `Phase` type union.
     */
    var PhaseKind;
    (function (PhaseKind) {
        PhaseKind[PhaseKind["Analysis"] = 0] = "Analysis";
        PhaseKind[PhaseKind["TypeCheckAndEmit"] = 1] = "TypeCheckAndEmit";
    })(PhaseKind || (PhaseKind = {}));
    /**
     * Manages the incremental portion of an Angular compilation, allowing for reuse of a prior
     * compilation if available, and producing an output state for reuse of the current compilation in a
     * future one.
     */
    var IncrementalCompilation = /** @class */ (function () {
        function IncrementalCompilation(state, depGraph, versions, step) {
            this.depGraph = depGraph;
            this.versions = versions;
            this.step = step;
            this._state = state;
            // The compilation begins in analysis phase.
            this.phase = {
                kind: PhaseKind.Analysis,
                semanticDepGraphUpdater: new semantic_graph_1.SemanticDepGraphUpdater(step !== null ? step.priorState.semanticDepGraph : null),
            };
        }
        /**
         * Begin a fresh `IncrementalCompilation`.
         */
        IncrementalCompilation.fresh = function (program, versions) {
            var state = {
                kind: state_1.IncrementalStateKind.Fresh,
            };
            return new IncrementalCompilation(state, new dependency_tracking_1.FileDependencyGraph(), versions, /* reuse */ null);
        };
        IncrementalCompilation.incremental = function (program, newVersions, oldProgram, oldState, modifiedResourceFiles, perf) {
            return perf.inPhase(perf_1.PerfPhase.Reconciliation, function () {
                var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e;
                var physicallyChangedTsFiles = new Set();
                var changedResourceFiles = new Set(modifiedResourceFiles !== null && modifiedResourceFiles !== void 0 ? modifiedResourceFiles : []);
                var priorAnalysis;
                switch (oldState.kind) {
                    case state_1.IncrementalStateKind.Fresh:
                        // Since this line of program has never been successfully analyzed to begin with, treat
                        // this as a fresh compilation.
                        return IncrementalCompilation.fresh(program, newVersions);
                    case state_1.IncrementalStateKind.Analyzed:
                        // The most recent program was analyzed successfully, so we can use that as our prior
                        // state and don't need to consider any other deltas except changes in the most recent
                        // program.
                        priorAnalysis = oldState;
                        break;
                    case state_1.IncrementalStateKind.Delta:
                        // There is an ancestor program which was analyzed successfully and can be used as a
                        // starting point, but we need to determine what's changed since that program.
                        priorAnalysis = oldState.lastAnalyzedState;
                        try {
                            for (var _f = tslib_1.__values(oldState.physicallyChangedTsFiles), _g = _f.next(); !_g.done; _g = _f.next()) {
                                var sfPath = _g.value;
                                physicallyChangedTsFiles.add(sfPath);
                            }
                        }
                        catch (e_1_1) { e_1 = { error: e_1_1 }; }
                        finally {
                            try {
                                if (_g && !_g.done && (_a = _f.return)) _a.call(_f);
                            }
                            finally { if (e_1) throw e_1.error; }
                        }
                        try {
                            for (var _h = tslib_1.__values(oldState.changedResourceFiles), _j = _h.next(); !_j.done; _j = _h.next()) {
                                var resourcePath = _j.value;
                                changedResourceFiles.add(resourcePath);
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_j && !_j.done && (_b = _h.return)) _b.call(_h);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                        break;
                }
                var oldVersions = priorAnalysis.versions;
                var oldFilesArray = oldProgram.getSourceFiles().map(toOriginalSourceFile);
                var oldFiles = new Set(oldFilesArray);
                var deletedTsFiles = new Set(oldFilesArray.map(function (sf) { return file_system_1.absoluteFromSourceFile(sf); }));
                try {
                    for (var _k = tslib_1.__values(program.getSourceFiles()), _l = _k.next(); !_l.done; _l = _k.next()) {
                        var possiblyRedirectedNewFile = _l.value;
                        var sf = toOriginalSourceFile(possiblyRedirectedNewFile);
                        var sfPath = file_system_1.absoluteFromSourceFile(sf);
                        // Since we're seeing a file in the incoming program with this name, it can't have been
                        // deleted.
                        deletedTsFiles.delete(sfPath);
                        if (oldFiles.has(sf)) {
                            // This source file has the same object identity as in the previous program. We need to
                            // determine if it's really the same file, or if it might have changed versions since the
                            // last program without changing its identity.
                            // If there's no version information available, then this is the same file, and we can
                            // skip it.
                            if (oldVersions === null || newVersions === null) {
                                continue;
                            }
                            // If a version is available for the file from both the prior and the current program, and
                            // that version is the same, then this is the same file, and we can skip it.
                            if (oldVersions.has(sfPath) && newVersions.has(sfPath) &&
                                oldVersions.get(sfPath) === newVersions.get(sfPath)) {
                                continue;
                            }
                            // Otherwise, assume that the file has changed. Either its versions didn't match, or we
                            // were missing version information about it on one side for some reason.
                        }
                        // Bail out if a .d.ts file changes - the semantic dep graph is not able to process such
                        // changes correctly yet.
                        if (sf.isDeclarationFile) {
                            return IncrementalCompilation.fresh(program, newVersions);
                        }
                        // The file has changed physically, so record it.
                        physicallyChangedTsFiles.add(sfPath);
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (_l && !_l.done && (_c = _k.return)) _c.call(_k);
                    }
                    finally { if (e_3) throw e_3.error; }
                }
                try {
                    // Remove any files that have been deleted from the list of physical changes.
                    for (var deletedTsFiles_1 = tslib_1.__values(deletedTsFiles), deletedTsFiles_1_1 = deletedTsFiles_1.next(); !deletedTsFiles_1_1.done; deletedTsFiles_1_1 = deletedTsFiles_1.next()) {
                        var deletedFileName = deletedTsFiles_1_1.value;
                        physicallyChangedTsFiles.delete(file_system_1.resolve(deletedFileName));
                    }
                }
                catch (e_4_1) { e_4 = { error: e_4_1 }; }
                finally {
                    try {
                        if (deletedTsFiles_1_1 && !deletedTsFiles_1_1.done && (_d = deletedTsFiles_1.return)) _d.call(deletedTsFiles_1);
                    }
                    finally { if (e_4) throw e_4.error; }
                }
                // Use the prior dependency graph to project physical changes into a set of logically changed
                // files.
                var depGraph = new dependency_tracking_1.FileDependencyGraph();
                var logicallyChangedTsFiles = depGraph.updateWithPhysicalChanges(priorAnalysis.depGraph, physicallyChangedTsFiles, deletedTsFiles, changedResourceFiles);
                try {
                    // Physically changed files aren't necessarily counted as logically changed by the dependency
                    // graph (files do not have edges to themselves), so add them to the logical changes
                    // explicitly.
                    for (var physicallyChangedTsFiles_1 = tslib_1.__values(physicallyChangedTsFiles), physicallyChangedTsFiles_1_1 = physicallyChangedTsFiles_1.next(); !physicallyChangedTsFiles_1_1.done; physicallyChangedTsFiles_1_1 = physicallyChangedTsFiles_1.next()) {
                        var sfPath = physicallyChangedTsFiles_1_1.value;
                        logicallyChangedTsFiles.add(sfPath);
                    }
                }
                catch (e_5_1) { e_5 = { error: e_5_1 }; }
                finally {
                    try {
                        if (physicallyChangedTsFiles_1_1 && !physicallyChangedTsFiles_1_1.done && (_e = physicallyChangedTsFiles_1.return)) _e.call(physicallyChangedTsFiles_1);
                    }
                    finally { if (e_5) throw e_5.error; }
                }
                // Start off in a `DeltaIncrementalState` as a delta against the previous successful analysis,
                // until this compilation completes its own analysis.
                var state = {
                    kind: state_1.IncrementalStateKind.Delta,
                    physicallyChangedTsFiles: physicallyChangedTsFiles,
                    changedResourceFiles: changedResourceFiles,
                    lastAnalyzedState: priorAnalysis,
                };
                return new IncrementalCompilation(state, depGraph, newVersions, {
                    priorState: priorAnalysis,
                    logicallyChangedTsFiles: logicallyChangedTsFiles,
                });
            });
        };
        Object.defineProperty(IncrementalCompilation.prototype, "state", {
            get: function () {
                return this._state;
            },
            enumerable: false,
            configurable: true
        });
        Object.defineProperty(IncrementalCompilation.prototype, "semanticDepGraphUpdater", {
            get: function () {
                if (this.phase.kind !== PhaseKind.Analysis) {
                    throw new Error("AssertionError: Cannot update the SemanticDepGraph after analysis completes");
                }
                return this.phase.semanticDepGraphUpdater;
            },
            enumerable: false,
            configurable: true
        });
        IncrementalCompilation.prototype.recordSuccessfulAnalysis = function (traitCompiler) {
            var e_6, _a, e_7, _b;
            if (this.phase.kind !== PhaseKind.Analysis) {
                throw new Error("AssertionError: Incremental compilation in phase " + PhaseKind[this.phase.kind] + ", expected Analysis");
            }
            var _c = this.phase.semanticDepGraphUpdater.finalize(), needsEmit = _c.needsEmit, needsTypeCheckEmit = _c.needsTypeCheckEmit, newGraph = _c.newGraph;
            // Determine the set of files which have already been emitted.
            var emitted;
            if (this.step === null) {
                // Since there is no prior compilation, no files have yet been emitted.
                emitted = new Set();
            }
            else {
                // Begin with the files emitted by the prior successful compilation, but remove those which we
                // know need to bee re-emitted.
                emitted = new Set(this.step.priorState.emitted);
                try {
                    // Files need re-emitted if they've logically changed.
                    for (var _d = tslib_1.__values(this.step.logicallyChangedTsFiles), _e = _d.next(); !_e.done; _e = _d.next()) {
                        var sfPath = _e.value;
                        emitted.delete(sfPath);
                    }
                }
                catch (e_6_1) { e_6 = { error: e_6_1 }; }
                finally {
                    try {
                        if (_e && !_e.done && (_a = _d.return)) _a.call(_d);
                    }
                    finally { if (e_6) throw e_6.error; }
                }
                try {
                    // Files need re-emitted if they've semantically changed.
                    for (var needsEmit_1 = tslib_1.__values(needsEmit), needsEmit_1_1 = needsEmit_1.next(); !needsEmit_1_1.done; needsEmit_1_1 = needsEmit_1.next()) {
                        var sfPath = needsEmit_1_1.value;
                        emitted.delete(sfPath);
                    }
                }
                catch (e_7_1) { e_7 = { error: e_7_1 }; }
                finally {
                    try {
                        if (needsEmit_1_1 && !needsEmit_1_1.done && (_b = needsEmit_1.return)) _b.call(needsEmit_1);
                    }
                    finally { if (e_7) throw e_7.error; }
                }
            }
            // Transition to a successfully analyzed compilation. At this point, a subsequent compilation
            // could use this state as a starting point.
            this._state = {
                kind: state_1.IncrementalStateKind.Analyzed,
                versions: this.versions,
                depGraph: this.depGraph,
                semanticDepGraph: newGraph,
                priorAnalysis: traitCompiler.getAnalyzedRecords(),
                typeCheckResults: null,
                emitted: emitted,
            };
            // We now enter the type-check and emit phase of compilation.
            this.phase = {
                kind: PhaseKind.TypeCheckAndEmit,
                needsEmit: needsEmit,
                needsTypeCheckEmit: needsTypeCheckEmit,
            };
        };
        IncrementalCompilation.prototype.recordSuccessfulTypeCheck = function (results) {
            if (this._state.kind !== state_1.IncrementalStateKind.Analyzed) {
                throw new Error("AssertionError: Expected successfully analyzed compilation.");
            }
            else if (this.phase.kind !== PhaseKind.TypeCheckAndEmit) {
                throw new Error("AssertionError: Incremental compilation in phase " + PhaseKind[this.phase.kind] + ", expected TypeCheck");
            }
            this._state.typeCheckResults = results;
        };
        IncrementalCompilation.prototype.recordSuccessfulEmit = function (sf) {
            if (this._state.kind !== state_1.IncrementalStateKind.Analyzed) {
                throw new Error("AssertionError: Expected successfully analyzed compilation.");
            }
            this._state.emitted.add(file_system_1.absoluteFromSourceFile(sf));
        };
        IncrementalCompilation.prototype.priorAnalysisFor = function (sf) {
            if (this.step === null) {
                return null;
            }
            var sfPath = file_system_1.absoluteFromSourceFile(sf);
            // If the file has logically changed, its previous analysis cannot be reused.
            if (this.step.logicallyChangedTsFiles.has(sfPath)) {
                return null;
            }
            var priorAnalysis = this.step.priorState.priorAnalysis;
            if (!priorAnalysis.has(sf)) {
                return null;
            }
            return priorAnalysis.get(sf);
        };
        IncrementalCompilation.prototype.priorTypeCheckingResultsFor = function (sf) {
            if (this.phase.kind !== PhaseKind.TypeCheckAndEmit) {
                throw new Error("AssertionError: Expected successfully analyzed compilation.");
            }
            if (this.step === null) {
                return null;
            }
            var sfPath = file_system_1.absoluteFromSourceFile(sf);
            // If the file has logically changed, or its template type-checking results have semantically
            // changed, then past type-checking results cannot be reused.
            if (this.step.logicallyChangedTsFiles.has(sfPath) ||
                this.phase.needsTypeCheckEmit.has(sfPath)) {
                return null;
            }
            // Past results also cannot be reused if they're not available.
            if (this.step.priorState.typeCheckResults === null ||
                !this.step.priorState.typeCheckResults.has(sfPath)) {
                return null;
            }
            var priorResults = this.step.priorState.typeCheckResults.get(sfPath);
            // If the past results relied on inlining, they're not safe for reuse.
            if (priorResults.hasInlines) {
                return null;
            }
            return priorResults;
        };
        IncrementalCompilation.prototype.safeToSkipEmit = function (sf) {
            // If this is a fresh compilation, it's never safe to skip an emit.
            if (this.step === null) {
                return false;
            }
            var sfPath = file_system_1.absoluteFromSourceFile(sf);
            // If the file has itself logically changed, it must be emitted.
            if (this.step.logicallyChangedTsFiles.has(sfPath)) {
                return false;
            }
            if (this.phase.kind !== PhaseKind.TypeCheckAndEmit) {
                throw new Error("AssertionError: Expected successful analysis before attempting to emit files");
            }
            // If during analysis it was determined that this file has semantically changed, it must be
            // emitted.
            if (this.phase.needsEmit.has(sfPath)) {
                return false;
            }
            // Generally it should be safe to assume here that the file was previously emitted by the last
            // successful compilation. However, as a defense-in-depth against incorrectness, we explicitly
            // check that the last emit included this file, and re-emit it otherwise.
            return this.step.priorState.emitted.has(sfPath);
        };
        return IncrementalCompilation;
    }());
    exports.IncrementalCompilation = IncrementalCompilation;
    /**
     * To accurately detect whether a source file was affected during an incremental rebuild, the
     * "original" source file needs to be consistently used.
     *
     * First, TypeScript may have created source file redirects when declaration files of the same
     * version of a library are included multiple times. The non-redirected source file should be used
     * to detect changes, as otherwise the redirected source files cause a mismatch when compared to
     * a prior program.
     *
     * Second, the program that is used for template type checking may contain mutated source files, if
     * inline type constructors or inline template type-check blocks had to be used. Such source files
     * store their original, non-mutated source file from the original program in a symbol. For
     * computing the affected files in an incremental build this original source file should be used, as
     * the mutated source file would always be considered affected.
     */
    function toOriginalSourceFile(sf) {
        var unredirectedSf = typescript_1.toUnredirectedSourceFile(sf);
        var originalFile = unredirectedSf[program_driver_1.NgOriginalFile];
        if (originalFile !== undefined) {
            return originalFile;
        }
        else {
            return unredirectedSf;
        }
    }
});
