// @angular/core/schematics/migrations/relative-link-resolution/collector.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/core/schematics/migrations/relative-link-resolution/collector", ["require", "exports", "typescript", "@angular/core/schematics/migrations/relative-link-resolution/util"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.RelativeLinkResolutionCollector = void 0;
    /**
     * @license
     * Copyright Google LLC All Rights Reserved.
     *
     * Use of this source code is governed by an MIT-style license that can be
     * found in the LICENSE file at https://angular.io/license
     */
    const ts = require("typescript");
    const util_1 = require("@angular/core/schematics/migrations/relative-link-resolution/util");
    /**
     * Visitor that walks through specified TypeScript nodes and collects all
     * found ExtraOptions#RelativeLinkResolution assignments.
     */
    class RelativeLinkResolutionCollector {
        constructor(typeChecker) {
            this.typeChecker = typeChecker;
            this.forRootCalls = [];
            this.extraOptionsLiterals = [];
        }
        visitNode(node) {
            let forRootCall = null;
            let literal = null;
            if (util_1.isRouterModuleForRoot(this.typeChecker, node) && node.arguments.length > 0) {
                if (node.arguments.length === 1) {
                    forRootCall = node;
                }
                else if (ts.isObjectLiteralExpression(node.arguments[1])) {
                    literal = node.arguments[1];
                }
                else if (ts.isIdentifier(node.arguments[1])) {
                    literal = this.getLiteralNeedingMigrationFromIdentifier(node.arguments[1]);
                }
            }
            else if (ts.isVariableDeclaration(node)) {
                literal = this.getLiteralNeedingMigration(node);
            }
            if (literal !== null) {
                this.extraOptionsLiterals.push(literal);
            }
            else if (forRootCall !== null) {
                this.forRootCalls.push(forRootCall);
            }
            else {
                // no match found, continue iteration
                ts.forEachChild(node, n => this.visitNode(n));
            }
        }
        getLiteralNeedingMigrationFromIdentifier(id) {
            const symbolForIdentifier = this.typeChecker.getSymbolAtLocation(id);
            if (symbolForIdentifier === undefined) {
                return null;
            }
            if (symbolForIdentifier.declarations === undefined ||
                symbolForIdentifier.declarations.length === 0) {
                return null;
            }
            const declarationNode = symbolForIdentifier.declarations[0];
            if (!ts.isVariableDeclaration(declarationNode) || declarationNode.initializer === undefined ||
                !ts.isObjectLiteralExpression(declarationNode.initializer)) {
                return null;
            }
            return declarationNode.initializer;
        }
        getLiteralNeedingMigration(node) {
            if (node.initializer === undefined) {
                return null;
            }
            // declaration could be `x: ExtraOptions = {}` or `x = {} as ExtraOptions`
            if (ts.isAsExpression(node.initializer) &&
                ts.isObjectLiteralExpression(node.initializer.expression) &&
                util_1.isExtraOptions(this.typeChecker, node.initializer.type)) {
                return node.initializer.expression;
            }
            else if (node.type !== undefined && ts.isObjectLiteralExpression(node.initializer) &&
                util_1.isExtraOptions(this.typeChecker, node.type)) {
                return node.initializer;
            }
            return null;
        }
    }
    exports.RelativeLinkResolutionCollector = RelativeLinkResolutionCollector;
});
