// @angular/compiler-cli/src/ngtsc/partial_evaluator/src/diagnostics.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/partial_evaluator/src/diagnostics", ["require", "exports", "tslib", "typescript", "@angular/compiler-cli/src/ngtsc/diagnostics", "@angular/compiler-cli/src/ngtsc/imports", "@angular/compiler-cli/src/ngtsc/partial_evaluator/src/dynamic", "@angular/compiler-cli/src/ngtsc/partial_evaluator/src/result"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.traceDynamicValue = exports.describeResolvedType = void 0;
    var tslib_1 = require("tslib");
    var ts = require("typescript");
    var diagnostics_1 = require("@angular/compiler-cli/src/ngtsc/diagnostics");
    var imports_1 = require("@angular/compiler-cli/src/ngtsc/imports");
    var dynamic_1 = require("@angular/compiler-cli/src/ngtsc/partial_evaluator/src/dynamic");
    var result_1 = require("@angular/compiler-cli/src/ngtsc/partial_evaluator/src/result");
    /**
     * Derives a type representation from a resolved value to be reported in a diagnostic.
     *
     * @param value The resolved value for which a type representation should be derived.
     * @param maxDepth The maximum nesting depth of objects and arrays, defaults to 1 level.
     */
    function describeResolvedType(value, maxDepth) {
        var _a, _b;
        if (maxDepth === void 0) { maxDepth = 1; }
        if (value === null) {
            return 'null';
        }
        else if (value === undefined) {
            return 'undefined';
        }
        else if (typeof value === 'number' || typeof value === 'boolean' || typeof value === 'string') {
            return typeof value;
        }
        else if (value instanceof Map) {
            if (maxDepth === 0) {
                return 'object';
            }
            var entries = Array.from(value.entries()).map(function (_a) {
                var _b = tslib_1.__read(_a, 2), key = _b[0], v = _b[1];
                return quoteKey(key) + ": " + describeResolvedType(v, maxDepth - 1);
            });
            return entries.length > 0 ? "{ " + entries.join('; ') + " }" : '{}';
        }
        else if (value instanceof result_1.ResolvedModule) {
            return '(module)';
        }
        else if (value instanceof result_1.EnumValue) {
            return (_a = value.enumRef.debugName) !== null && _a !== void 0 ? _a : '(anonymous)';
        }
        else if (value instanceof imports_1.Reference) {
            return (_b = value.debugName) !== null && _b !== void 0 ? _b : '(anonymous)';
        }
        else if (Array.isArray(value)) {
            if (maxDepth === 0) {
                return 'Array';
            }
            return "[" + value.map(function (v) { return describeResolvedType(v, maxDepth - 1); }).join(', ') + "]";
        }
        else if (value instanceof dynamic_1.DynamicValue) {
            return '(not statically analyzable)';
        }
        else if (value instanceof result_1.KnownFn) {
            return 'Function';
        }
        else {
            return 'unknown';
        }
    }
    exports.describeResolvedType = describeResolvedType;
    function quoteKey(key) {
        if (/^[a-z0-9_]+$/i.test(key)) {
            return key;
        }
        else {
            return "'" + key.replace(/'/g, '\\\'') + "'";
        }
    }
    /**
     * Creates an array of related information diagnostics for a `DynamicValue` that describe the trace
     * of why an expression was evaluated as dynamic.
     *
     * @param node The node for which a `ts.Diagnostic` is to be created with the trace.
     * @param value The dynamic value for which a trace should be created.
     */
    function traceDynamicValue(node, value) {
        return value.accept(new TraceDynamicValueVisitor(node));
    }
    exports.traceDynamicValue = traceDynamicValue;
    var TraceDynamicValueVisitor = /** @class */ (function () {
        function TraceDynamicValueVisitor(node) {
            this.node = node;
            this.currentContainerNode = null;
        }
        TraceDynamicValueVisitor.prototype.visitDynamicInput = function (value) {
            var trace = value.reason.accept(this);
            if (this.shouldTrace(value.node)) {
                var info = diagnostics_1.makeRelatedInformation(value.node, 'Unable to evaluate this expression statically.');
                trace.unshift(info);
            }
            return trace;
        };
        TraceDynamicValueVisitor.prototype.visitDynamicString = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'A string value could not be determined statically.')];
        };
        TraceDynamicValueVisitor.prototype.visitExternalReference = function (value) {
            var name = value.reason.debugName;
            var description = name !== null ? "'" + name + "'" : 'an anonymous declaration';
            return [diagnostics_1.makeRelatedInformation(value.node, "A value for " + description + " cannot be determined statically, as it is an external declaration.")];
        };
        TraceDynamicValueVisitor.prototype.visitComplexFunctionCall = function (value) {
            return [
                diagnostics_1.makeRelatedInformation(value.node, 'Unable to evaluate function call of complex function. A function must have exactly one return statement.'),
                diagnostics_1.makeRelatedInformation(value.reason.node, 'Function is declared here.')
            ];
        };
        TraceDynamicValueVisitor.prototype.visitInvalidExpressionType = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'Unable to evaluate an invalid expression.')];
        };
        TraceDynamicValueVisitor.prototype.visitUnknown = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'Unable to evaluate statically.')];
        };
        TraceDynamicValueVisitor.prototype.visitUnknownIdentifier = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'Unknown reference.')];
        };
        TraceDynamicValueVisitor.prototype.visitDynamicType = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'Dynamic type.')];
        };
        TraceDynamicValueVisitor.prototype.visitUnsupportedSyntax = function (value) {
            return [diagnostics_1.makeRelatedInformation(value.node, 'This syntax is not supported.')];
        };
        /**
         * Determines whether the dynamic value reported for the node should be traced, i.e. if it is not
         * part of the container for which the most recent trace was created.
         */
        TraceDynamicValueVisitor.prototype.shouldTrace = function (node) {
            if (node === this.node) {
                // Do not include a dynamic value for the origin node, as the main diagnostic is already
                // reported on that node.
                return false;
            }
            var container = getContainerNode(node);
            if (container === this.currentContainerNode) {
                // The node is part of the same container as the previous trace entry, so this dynamic value
                // should not become part of the trace.
                return false;
            }
            this.currentContainerNode = container;
            return true;
        };
        return TraceDynamicValueVisitor;
    }());
    /**
     * Determines the closest parent node that is to be considered as container, which is used to reduce
     * the granularity of tracing the dynamic values to a single entry per container. Currently, full
     * statements and destructuring patterns are considered as container.
     */
    function getContainerNode(node) {
        var currentNode = node;
        while (currentNode !== undefined) {
            switch (currentNode.kind) {
                case ts.SyntaxKind.ExpressionStatement:
                case ts.SyntaxKind.VariableStatement:
                case ts.SyntaxKind.ReturnStatement:
                case ts.SyntaxKind.IfStatement:
                case ts.SyntaxKind.SwitchStatement:
                case ts.SyntaxKind.DoStatement:
                case ts.SyntaxKind.WhileStatement:
                case ts.SyntaxKind.ForStatement:
                case ts.SyntaxKind.ForInStatement:
                case ts.SyntaxKind.ForOfStatement:
                case ts.SyntaxKind.ContinueStatement:
                case ts.SyntaxKind.BreakStatement:
                case ts.SyntaxKind.ThrowStatement:
                case ts.SyntaxKind.ObjectBindingPattern:
                case ts.SyntaxKind.ArrayBindingPattern:
                    return currentNode;
            }
            currentNode = currentNode.parent;
        }
        return node.getSourceFile();
    }
});
